File: sdks/node/src/runtime/acknowledgements.ts
import {
  runtimeConfigurationAcknowledgementSchema,
  type RuntimeConfigurationAcknowledgement,
} from "@tokenpilot/contracts";
import { ulid } from "ulid";

import { AiControlSdkError } from "../errors.js";

export type RuntimeAcknowledgementState = "received" | "applied" | "rejected";

export interface RuntimeSdkIdentity {
  readonly instanceId: string;
  readonly version: string;
}

function record(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null ? (value as Record<string, unknown>) : null;
}

function configurationReferences(value: unknown): readonly {
  readonly applicationId: string;
  readonly configurationVersion: number;
  readonly configurationEtag: string;
}[] {
  const snapshot = record(value);
  const routing = snapshot?.routing;
  const plans = record(routing);
  if (plans === null) return [];
  const versions = new Set<number>();
  for (const rawPlan of Object.values(plans)) {
    const plan = record(rawPlan);
    const version = plan?.configuration_version;
    if (typeof version === "number" && Number.isSafeInteger(version) && version > 0) {
      versions.add(version);
    }
  }
  const [configurationVersion] = [...versions];
  const configurationEtag = snapshot?.etag;
  const applicationId = snapshot?.application_id;
  if (
    versions.size !== 1 ||
    configurationVersion === undefined ||
    typeof applicationId !== "string" ||
    !/^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/iu.test(
      applicationId,
    ) ||
    typeof configurationEtag !== "string" ||
    !/^sha256:[0-9a-f]{64}$/u.test(configurationEtag)
  ) {
    return [];
  }
  return [{ applicationId, configurationVersion, configurationEtag }];
}

export function runtimeAcknowledgements(
  value: unknown,
  state: RuntimeAcknowledgementState,
  identity: RuntimeSdkIdentity,
  now: Date,
  error?: Error,
): readonly RuntimeConfigurationAcknowledgement[] {
  return configurationReferences(value).map(
    ({ applicationId, configurationVersion, configurationEtag }) =>
      runtimeConfigurationAcknowledgementSchema.parse({
        schema_version: "2.0",
        application_id: applicationId,
        acknowledgement_id: ulid(now.getTime()),
        acknowledged_at: now.toISOString(),
        connector: { instance_id: identity.instanceId, name: "node", version: identity.version },
        configuration_version: configurationVersion,
        configuration_etag: configurationEtag,
        state,
        applied_at: state === "applied" ? now.toISOString() : null,
        error:
          state === "rejected"
            ? {
                code: "SDK_RUNTIME_SNAPSHOT_REJECTED",
                message: (error?.message ?? "Runtime Snapshot was rejected").slice(0, 500),
              }
            : null,
      }),
  );
}

function errorValue(value: unknown): Error {
  return value instanceof Error ? value : new Error("Unknown acknowledgement failure");
}

export class RuntimeAcknowledgementQueue {
  readonly #controlPlaneUrl: string;
  readonly #apiKey: string;
  readonly #identity: RuntimeSdkIdentity;
  readonly #fetch: typeof fetch;
  readonly #now: () => Date;
  readonly #onError: (error: Error) => void;
  readonly #pending: RuntimeConfigurationAcknowledgement[] = [];

  public constructor(options: {
    readonly controlPlaneUrl: string;
    readonly apiKey: string;
    readonly identity: RuntimeSdkIdentity;
    readonly fetch: typeof fetch;
    readonly now: () => Date;
    readonly onError: (error: Error) => void;
  }) {
    if (options.identity.instanceId.length < 1 || options.identity.instanceId.length > 256) {
      throw new AiControlSdkError(
        "SDK_INVALID_CONFIGURATION",
        "SDK instanceId must contain between 1 and 256 characters.",
      );
    }
    if (!/^\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?$/u.test(options.identity.version)) {
      throw new AiControlSdkError("SDK_INVALID_CONFIGURATION", "sdkVersion must be semantic.");
    }
    this.#controlPlaneUrl = options.controlPlaneUrl;
    this.#apiKey = options.apiKey;
    this.#identity = options.identity;
    this.#fetch = options.fetch;
    this.#now = options.now;
    this.#onError = options.onError;
  }

  public queue(value: unknown, state: RuntimeAcknowledgementState, error?: Error): void {
    this.#pending.push(
      ...runtimeAcknowledgements(value, state, this.#identity, this.#now(), error),
    );
  }

  public async flush(required: boolean): Promise<void> {
    while (this.#pending.length > 0) {
      try {
        const response = await this.#fetch(
          `${this.#controlPlaneUrl}/runtime/configuration-acknowledgements`,
          {
            method: "POST",
            headers: {
              authorization: `Bearer ${this.#apiKey}`,
              "content-type": "application/json",
            },
            body: JSON.stringify(this.#pending[0]),
          },
        );
        if (!response.ok) {
          throw new AiControlSdkError(
            "SDK_RUNTIME_ACK_FAILED",
            `Control Plane rejected runtime configuration acknowledgement with HTTP ${response.status}.`,
          );
        }
        this.#pending.shift();
      } catch (error) {
        const failure = errorValue(error);
        if (required) throw failure;
        this.#onError(failure);
        return;
      }
    }
  }
}

File: sdks/python/src/ai_control_sdk/runtime/acknowledgements.py
"""Current application configuration acknowledgements emitted by Python clients."""

from __future__ import annotations

import re
from collections.abc import Mapping
from datetime import datetime
from typing import Any, Literal

from .context import new_ulid
from .contracts import (
    ETAG_PATTERN,
    UUID_PATTERN,
    RuntimeAcknowledgementError,
    RuntimeConfigurationAcknowledgement,
    RuntimeConnectorIdentity,
    RuntimeSnapshot,
)
from .state import utc_string

AcknowledgementState = Literal["received", "applied", "rejected"]


def _mapping(value: object) -> Mapping[str, Any] | None:
    return value if isinstance(value, Mapping) else None


def _configuration_reference(value: object) -> tuple[str, int, str] | None:
    raw: object = (
        value.model_dump(mode="json", by_alias=True)
        if isinstance(value, RuntimeSnapshot)
        else value
    )
    snapshot = _mapping(raw)
    plans = _mapping(snapshot.get("routing")) if snapshot is not None else None
    if plans is None:
        return None
    versions = {
        plan.get("configuration_version")
        for raw_plan in plans.values()
        if (plan := _mapping(raw_plan)) is not None
        and isinstance(plan.get("configuration_version"), int)
        and not isinstance(plan.get("configuration_version"), bool)
    }
    etag = snapshot.get("etag") if snapshot is not None else None
    application_id = snapshot.get("application_id") if snapshot is not None else None
    if (
        len(versions) != 1
        or not isinstance(application_id, str)
        or re.fullmatch(UUID_PATTERN, application_id) is None
        or not isinstance(etag, str)
        or re.fullmatch(ETAG_PATTERN, etag) is None
    ):
        return None
    version = next(iter(versions))
    return (application_id, version, etag) if isinstance(version, int) and version > 0 else None


def runtime_acknowledgements(
    value: object,
    state: AcknowledgementState,
    identity: RuntimeConnectorIdentity,
    now: datetime,
    error: Exception | None = None,
) -> list[RuntimeConfigurationAcknowledgement]:
    reference = _configuration_reference(value)
    if reference is None:
        return []
    application_id, version, etag = reference
    timestamp = utc_string(now)
    return [
        RuntimeConfigurationAcknowledgement(
            schema_version="2.0",
            application_id=application_id,
            acknowledgement_id=new_ulid().upper(),
            acknowledged_at=timestamp,
            connector=identity,
            configuration_version=version,
            configuration_etag=etag,
            state=state,
            applied_at=timestamp if state == "applied" else None,
            error=(
                RuntimeAcknowledgementError(
                    code="SDK_RUNTIME_SNAPSHOT_REJECTED",
                    message=str(error or "Runtime Snapshot was rejected")[:500],
                )
                if state == "rejected"
                else None
            ),
        )
    ]
